src/odemis/acq/align/autofocus.py

```diff
@@ -120,21 +121,31 @@ def Measure1d(image):
     Given an image of a 1 line ccd, measure the focus based on the inverse of the width of a gaussian fit of the data.
     It is assumed that the signal is in focus when the width of the signal is smallest and therefore sigma is smallest.
     image (model.DataArray): 1D image from 1 line ccd.
-    returns (float): The focus level of the image.
+    returns (float): The focus level of the image, based on the inverse of the width of a gaussian fitted on the image.
     """
-    def gauss(x, *p):
-        a, b, c, d = p
-        y = a * numpy.exp(-numpy.power((x - b), 2.) / (2. * c ** 2.)) + d
+    # Use the gauss function to fit a gaussian to the 1 line image.
+    def gauss(x, amplitude, pos, width, base):
+        y = amplitude * numpy.exp(-(x - pos) ** 2 / (2 * width ** 2)) + base
         return y
     # squeeze to make sure the image array is 1d.
     signal = numpy.squeeze(image)
+    # Apply a median filter with a kernel of 5, to filter out (at the most 2 neighbouring) outliers.
+    signal = medfilt(signal, 5)
     x = numpy.arange(len(signal))
-    # give an initial estimate for the parameters of the gaussian fit: [max, mu, sigma, y-offset]
-    p_initial = [max(signal), numpy.argmax(signal), 10, min(signal)]
+    # determine the indices and the values of the 1% highest points in the signal.
+    max_ids = signal.argsort()[-int(0.01 * len(signal)):]
+    max_sig = signal[max_ids]
+    # give an initial estimate for the parameters of the gaussian fit: [amplitude, expected position, width, base]
+    p_initial = [numpy.median(max_sig) - numpy.median(signal), numpy.median(max_ids), 0.01 * len(signal),
+                 numpy.median(signal)]
     # Use curve_fit to fit the gauss function to the data. Use p_initial as our initial guess.
     popt, pcov = curve_fit(gauss, x, signal, p0=p_initial)
-    # The focus metric is the inverse of sigma of the gaussian fit; a smaller sigma is a higher focus level.
-    return 1 / abs(popt[2])
+    # Determine the goodness of the fit. It is expected that in the area of focus the goodness of fit is high.
+    y_fit = gauss(x, *popt)
+    r2 = 1 - (((signal - y_fit) ** 2).sum() / ((signal - signal.mean()) ** 2).sum())
+    # The focus metric is a combination of the inverse of sigma of the gaussian fit (a smaller sigma is a higher focus
```

Comment:
Have you tried using the amplitude (compared to the base)? It's very typical that focused image have brighter pixels, so it should be a quite reliable indicator.
In addition, I'm a little worried by the gaussian fit, because I wonder how well it performs if the signal happens to not really be a gaussian (eg, a square wave).

Reply:
The problem I'm seeing with using the amplitude (for the simulated focus test) is that there is a spread around the focus point where the amplitude is equal to that of the amplitude in the focus point. 

I tried combining the amplitude with the width, however I can't think of a way to scale them so I can add them together since the amplitude has a much higher value than the width.

I also tried what you suggested in the email to take the sum of the X% highest pixels. That has the following problem: if the in focus peak has a very small width and you go slightly out of focus there are more bright pixels thus a the sum of the X% highest pixels is higher when slightly out of focus then when in focus.

I agree that the goodness of fit is probably not the best metric, especially for a square wave.